### PR TITLE
[E129-MG < T1003-MG] Expand fine grained access checker with more granular permissions

### PR DESCRIPTION
**Pull request by BorisTasevski**

### CODE REVIEW

File: src/auth/models.cpp [hunk 1]
```diff
@@ -282,8 +305,34 @@ FineGrainedAccessPermissions FineGrainedAccessPermissions::Deserialize(const nlo
...
+
+  while (u_label_perm != 0) {
+    perm |= u_label_perm;
+    u_label_perm >>= shift;
+  }
+
+  grants_[permission] = perm;
+}
+
+void FineGrainedAccessPermissions::deny_(const std::string &permission, LabelPermission label_permission) {
```
Comment (Josipmrden) [inline src/auth/models.cpp:324]: is this common to have underscore at the end

File: src/auth/models.cpp [hunk 2]
```diff
@@ -282,8 +305,34 @@ FineGrainedAccessPermissions FineGrainedAccessPermissions::Deserialize(const nlo
...
+  uint64_t shift = 1;
+  uint64_t perm = 0;
+  auto u_label_perm = static_cast<uint64_t>(label_permission);
```
Comment (Josipmrden) [inline src/auth/models.cpp:327]: do we need 64 bits

File: tests/e2e/write_procedures/procedures/read.py [hunk 3]
```diff
@@ -13,8 +13,7 @@
 
 
 @mgp.read_proc
-def underlying_graph_is_mutable(ctx: mgp.ProcCtx,
-                                object: mgp.Any) -> mgp.Record(mutable=bool):
+def underlying_graph_is_mutable(ctx: mgp.ProcCtx, object: mgp.Any) -> mgp.Record(mutable=bool):
```
Comment (Josipmrden) [inline tests/e2e/write_procedures/procedures/read.py:16]: why is this changed? only formatting?

File: src/auth/models.cpp (hunk 1 above)
Comment (BorisTasevski) [reply src/auth/models.cpp]: No naming fixes are already made

File: src/auth/models.cpp (hunk 2 above)
Comment (BorisTasevski) [reply src/auth/models.cpp:315]: For now, yes. The real reason is bellow, shift operations convert operators to uint64_t, hence the 64bits, not to get lost in conversions

File: tests/e2e/write_procedures/procedures/read.py (hunk 3 above)
Comment (BorisTasevski) [reply tests/e2e/write_procedures/procedures/read.py:16]: Yes, slipped me...

File: src/auth/models.cpp [hunk 4]
```diff
@@ -98,6 +104,19 @@ std::string PermissionLevelToString(PermissionLevel level) {
...
+
+  for (const auto &it : second) {
+    if (result.find(it.first) != result.end()) {
+      result[it.first] |= it.second;
+    }
+  }
+
+  return result;
+}
+
```
Comment (Josipmrden) [inline src/auth/models.cpp:119]: Can you please provide documentation what you're doing here, I always forget

File: src/auth/models.cpp [hunk 5]
```diff
@@ -183,45 +202,44 @@ bool operator==(const Permissions &first, const Permissions &second) {
...
+  const auto &deny_it = DeniesFind(permission);
+
+  if (deny_it != denies_.end() && (deny_it->second & label_permission)) {
+    denies_permission = PermissionLevel::DENY;
   }
 
-  if ((grants_.size() == 1 && grants_.find(ASTERISK) != grants_.end()) || grants_.find(permission) != denies_.end()) {
-    return PermissionLevel::GRANT;
+  const auto &grant_it = GrantsFind(permission);
```
Comment (Josipmrden) [inline src/auth/models.cpp:220]: FindGrants?

File: tests/unit/auth.cpp [hunk 6]
```diff
@@ -407,6 +407,286 @@ TEST(AuthWithoutStorage, PermissionsMaskTest) {
...
+    FineGrainedAccessPermissions fga_permissions1, fga_permissions2;
+    ASSERT_TRUE(fga_permissions1 == fga_permissions2);
+  }
+
+  {
+    FineGrainedAccessPermissions fga_permissions;
+    ASSERT_TRUE(fga_permissions.grants().empty());
+    ASSERT_TRUE(fga_permissions.denies().empty());
+
+    ASSERT_EQ(fga_permissions.Has("AnyString", LabelPermission::CREATE_DELETE), PermissionLevel::DENY);
```
Comment (Josipmrden) [inline tests/unit/auth.cpp:421]: put "AnyString" in a var

File: tests/unit/auth.cpp [hunk 7]
```diff
@@ -407,6 +407,286 @@ TEST(AuthWithoutStorage, PermissionsMaskTest) {
...
+    ASSERT_EQ(fga_permissions.Has("AnyString", LabelPermission::READ), PermissionLevel::DENY);
+  }
+
+  {
+    FineGrainedAccessPermissions fga_permissions;
+    fga_permissions.Deny("*", LabelPermission::READ);
+
+    ASSERT_EQ(fga_permissions.Has("AnyString", LabelPermission::CREATE_DELETE), PermissionLevel::DENY);
+    ASSERT_EQ(fga_permissions.Has("AnyString", LabelPermission::EDIT), PermissionLevel::DENY);
+    ASSERT_EQ(fga_permissions.Has("AnyString", LabelPermission::READ), PermissionLevel::DENY);
```
Comment (Josipmrden) [inline tests/unit/auth.cpp:566]: this test is the same as the test (2 above). maybe you meant create delete instead of read

File: tests/unit/auth.cpp [hunk 8]
```diff
@@ -407,6 +407,286 @@ TEST(AuthWithoutStorage, PermissionsMaskTest) {
...
+  }
+
+  {
+    FineGrainedAccessPermissions fga_permissions;
+    fga_permissions.Grant("*", LabelPermission::CREATE_DELETE);
```
Comment (Josipmrden) [inline tests/unit/auth.cpp:571]: what do you think of naming WRITE instead of CREATE_DELETE?

Comment (Josipmrden) [reply tests/unit/auth.cpp:571]: or MANAGE, or something similar?

Comment (BorisTasevski) [reply tests/unit/auth.cpp:571]: We can discuss on it, Yes

File: src/auth/models.cpp (hunk 5 above)
Comment (BorisTasevski) [reply src/auth/models.cpp:220]: FindInGrants

**Review by Josipmrden — APPROVED** [review]

File: src/auth/models.cpp [hunk 9]
```diff
@@ -101,6 +100,24 @@ std::string PermissionLevelToString(PermissionLevel level) {
   }
 }
 
+FineGrainedAccessPermissions Merge(const FineGrainedAccessPermissions &first,
+                                   const FineGrainedAccessPermissions &second) {
+  std::unordered_map<std::string, uint64_t> permissions{first.permissions()};
+  std::optional<uint64_t> global_permission = std::nullopt;
```
Comment (jbajic) [inline src/auth/models.cpp:106]: No need for explicit initialization, it is implicitly initialized to nullopt

File: src/auth/models.cpp [hunk 10]
```diff
@@ -101,6 +100,24 @@ std::string PermissionLevelToString(PermissionLevel level) {
...
+
+  if (second.global_permission().has_value()) {
+    global_permission = second.global_permission().value();
+  } else if (first.global_permission().has_value()) {
+    global_permission = first.global_permission().value();
+  }
```
Comment (jbajic) [inline src/auth/models.cpp:112]: Hmm I don't quite understand this logic, what if first and second have values you would just ignore values in first `FineGrainedAccessPermissions`, and why?

File: src/auth/models.cpp [hunk 11]
```diff
@@ -186,112 +203,102 @@ bool operator==(const Permissions &first, const Permissions &second) {
...
 
-PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission) const {
-  if ((denies_.size() == 1 && denies_.find(ASTERISK) != denies_.end()) || denies_.find(permission) != denies_.end()) {
-    return PermissionLevel::DENY;
-  }
+PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission,
+                                                  const LabelPermission label_permission) {
+  uint64_t concrete_permission = permissions_.contains(permission)
+                                     ? permissions_[permission]
+                                     : (global_permission_.has_value() ? global_permission_.value() : 0);
```
Comment (jbajic) [inline src/auth/models.cpp:214]: I think this is more readable in my opinion, multiple ternary operators are a bit confusing, definitely include `<functional>` for invoke
```suggestion
  const auto concrete_permission = std::invoke([&]() -> uint64_t {
    if (permissions_.contains(permission)) {
      return permissions_[permission];
    }
    if (global_permission_.has_value()) {
      return global_permission_.value();
    }
    return 0;
  });
```

File: src/auth/models.cpp [hunk 12]
```diff
@@ -186,112 +203,102 @@ bool operator==(const Permissions &first, const Permissions &second) {
...
 
-  if ((grants_.size() == 1 && grants_.find(ASTERISK) != grants_.end()) || grants_.find(permission) != denies_.end()) {
-    return PermissionLevel::GRANT;
-  }
+  auto temp_permission = concrete_permission & label_permission;
```
Comment (jbajic) [inline src/auth/models.cpp:216]:
```suggestion
  const auto temp_permission = concrete_permission & label_permission;
```

File: src/memgraph.cpp [hunk 13]
```diff
@@ -756,43 +756,56 @@ class AuthQueryHandler final : public memgraph::query::AuthQueryHandler {
...
-    EditPermissions(user_or_role, privileges, labels, edgeTypes, [](auto *permissions, const auto &permission) {
-      // TODO (mferencevic): should we first check that the
-      // privilege is granted/denied/revoked before
-      // unconditionally granting/denying/revoking it?
-      permissions->Grant(permission);
-    });
+                      const std::vector<std::string> &labels, const std::vector<std::string> &edge_types) override {
+    EditPermissions(
+        user_or_role, privileges, labels, edge_types,
+        [](auto *permissions, const auto &permission) {
```
Comment (jbajic) [inline src/memgraph.cpp:762]: How hard is it t change this pointer to reference?

File: tests/unit/auth.cpp [hunk 14]
```diff
@@ -475,6 +484,360 @@ TEST(AuthWithoutStorage, PermissionsMaskTest) {
...
+    fga_permissions2.Grant(check_label, LabelPermission::READ);
+
+    auto fga_permissions3 = memgraph::auth::Merge(fga_permissions1, fga_permissions2);
+
+    ASSERT_EQ(fga_permissions3.Has(check_label, LabelPermission::CREATE_DELETE), PermissionLevel::DENY);
+    ASSERT_EQ(fga_permissions3.Has(check_label, LabelPermission::EDIT), PermissionLevel::DENY);
+    ASSERT_EQ(fga_permissions3.Has(check_label, LabelPermission::READ), PermissionLevel::GRANT);
+  }
+}
 TEST(AuthWithoutStorage, UserSerializeDeserialize) {
```
Comment (jbajic) [inline tests/unit/auth.cpp:841]:
```suggestion
}

TEST(AuthWithoutStorage, UserSerializeDeserialize) {
```

File: src/auth/models.hpp [hunk 15]
```diff
@@ -44,15 +44,32 @@ enum class Permission : uint64_t {
 };
 // clang-format on
 
+// clang-format off
+enum class LabelPermission : uint64_t {
+  READ          = 1,
+  EDIT          = 1U << 1U,
```
[the remark this answers is not in the document]
Comment (BorisTasevski) [reply src/auth/models.hpp:50]: Whole naming will be addressed in the following PRs, discussion pending

File: src/auth/models.cpp [hunk 16]
```diff
@@ -101,6 +100,24 @@ std::string PermissionLevelToString(PermissionLevel level) {
...
+
+  if (second.global_permission().has_value()) {
+    global_permission = second.global_permission().value();
+  } else if (first.global_permission().has_value()) {
+    global_permission = first.global_permission().value();
+  }
+
+  for (const auto &it : second.permissions()) {
+    permissions[it.first] = it.second;
+  }
```
Comment (jbajic) [inline src/auth/models.cpp:116]:
```suggestion
  for (const auto &[label_name, permission] : second.permissions()) {
    permissions[label_name] = permission;
  }
```

File: src/auth/models.cpp (hunk 10 above)
Comment (jbajic) [reply src/auth/models.cpp:112]: Closing this comment, was resolved in private discussion between me and Boris

File: src/auth/models.cpp (hunk 16 above)
Comment (BorisTasevski) [reply src/auth/models.cpp]: Nice, i forgot about this

File: src/auth/models.cpp [hunk 17]
```diff
@@ -186,112 +203,111 @@ bool operator==(const Permissions &first, const Permissions &second) {
...
-const std::string ASTERISK = "*";
+FineGrainedAccessPermissions::FineGrainedAccessPermissions(const std::unordered_map<std::string, uint64_t> &permissions,
+                                                           const std::optional<uint64_t> &global_permission)
+    : permissions_(permissions), global_permission_(global_permission) {}
+
+PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission,
+                                                  const LabelPermission label_permission) const {
+  const auto concrete_permission = std::invoke([&]() -> uint64_t {
+    if (permissions_.contains(permission)) {
+      return permissions_.at(permission);
```
Comment (jbajic) [inline src/auth/models.cpp:214]: No need for at since at throws exception, but you have verified in if that there definitely is a permisson
```suggestion
      return permissions_[permission];
```

File: src/auth/models.hpp [hunk 18]
```diff
@@ -44,15 +44,34 @@ enum class Permission : uint64_t {
...
 // Function that converts a permission to its string representation.
 std::string PermissionToString(Permission permission);
 
 // Class that indicates what permission level the user/role has.
-enum class PermissionLevel {
-  GRANT,
-  NEUTRAL,
-  DENY,
-};
+enum class PermissionLevel : short { GRANT, NEUTRAL, DENY };
```
Comment (jbajic) [inline src/auth/models.hpp:74]: I would rather be specific on the type, maybe a suggestion suits you
```suggestion
enum class PermissionLevel : uint8_t { GRANT, NEUTRAL, DENY };
```

File: src/memgraph.cpp [hunk 19]
```diff
@@ -756,43 +756,56 @@ class AuthQueryHandler final : public memgraph::query::AuthQueryHandler {
...
+          // privilege is granted/denied/revoked before
+          // unconditionally granting/denying/revoking it?
+          permissions.Revoke(permission);
+        },
+        [](auto &label_permissions, const auto &label) { label_permissions.Revoke(label); });
   }
 
  private:
-  template <class TEditFun>
+  template <class TEditFun, class TEditLabelPermisionsFun>
```
Comment (jbajic) [inline src/memgraph.cpp:804]:
```suggestion
  template <class TEditFun, class TEditLabelPermissionsFun>
```

File: src/memgraph.cpp [hunk 20]
```diff
@@ -756,43 +756,56 @@ class AuthQueryHandler final : public memgraph::query::AuthQueryHandler {
...
   void EditPermissions(const std::string &user_or_role,
                        const std::vector<memgraph::query::AuthQuery::Privilege> &privileges,
                        const std::vector<std::string> &labels, const std::vector<std::string> &edgeTypes,
-                       const TEditFun &edit_fun) {
+                       const TEditFun &edit_fun, const TEditLabelPermisionsFun &edit_label_permisions_fun) {
```
Comment (jbajic) [inline src/memgraph.cpp:808]: The typo is also below
```suggestion
                       const TEditFun &edit_fun, const TEditLabelPermisionsFun &edit_label_permissions_fun) {
```

File: src/auth/models.cpp (hunk 17 above)
Comment (BorisTasevski) [reply src/auth/models.cpp:214]: Yes, but the method cant be marked as const then

**Review by jbajic — APPROVED** [review]
